src/alchemlyb/parsing/amber.py

```diff
@@ -122,18 +122,19 @@ def extract_section(self, start, end, fields, limit=None, extra='',
         line = ''.join(lines)
         result = []
         for field in fields:
-            match = re.search(r' %s\s+=\s+(\*+|%s|\d+)'
-                              % (field, _FP_RE), line)
+            match = re.search(fr' {field}\s+=\s+(\*+|{_FP_RE}|\d+)', line)
             if match:
                 value = match.group(1)
-                # FIXME: assumes fields are only integers or floats
-                if '*' in value:  # Fortran format overflow
+                if '*' in value:  # catch fortran format overflow
                     result.append(float('Inf'))
-                # NOTE: check if this is a sufficient test for int
-                elif '.' not in value and re.search(r'\d+', value):
-                    result.append(int(value))
                 else:
-                    result.append(float(value))
+                    try:
+                        result.append(int(value))
+                    except ValueError:
+                        try:
+                            result.append(float(value))
+                        except ValueError:
+                            result.append(str(value))
```

Comment:
This section is inside the class `SectionParser`, particularly this method is called `extract_section`. 
It is very general and is able to extract data values in fields from a section of a file, and the section is marked with start and end regexes given in input. So this method is indeed very general.
I think we have two options here:

1. keep the method as general as possible, and leave the responsibility to the user of the method to check that it's extracting the right types.
2. make the method "tailor-made" for us, and raise an exception right here if a `str` is found.

Maybe 2. it's the better choice here, to fail early and quickly if problem arises.

Reply:
On second thought, this method is used to parse AMBER file and parse the different sections, reading the different options used during the simulation. Only a small part of the data extracted with this method will end up into a DataFrame, the bigger part is just `option1=val1, option2=val2...`.
So keeping the option to read a string should be advisable.

Reply:
Is there a use case where the original parser cannot parse the amber out file but the new parser could do?
My concern is that the original parser will not output a `string`, if the content is not `int` or `float`, it will be cast as `None` and it works fine so far (we didn't get complaints about it cannot parse amber output). So casting anything that cannot be cast as `int` or `float` to `string` makes me worry that it will mask some issue that should be flagged.

Reply:
Currently, if it would ever find a string it would not output `None`, simply result would be an empty list, and then it's up to the unpacking of the results from the method call that should handle the case (we don't currently have test-cases for that, but we don't have any part of the code that has this kind of problem, either).

None is returned if nothing is matched by the regex!

Reply:
I think I don't like that the thing which cannot be parsed by int/float will be cast as a string. This could introduce errors where incorrectly formatted numbers were cast as strings.
I think unless there is a compelling reason such as the current parser cannot parse a normal AMBER file, I don't want to boarden the scoop of the current parser.

By the way, the patch coverage is a bit low for the PR, we are fine with the coverage of the project going down but would like the patch coverage to be high.

Reply:
I just removed the try/except block that parsed the string, in this way the coverage is back to normal and we avoid the issue of casting incorrectly formatted numbers to strings! If we'll have to make some more fancy parsing of the AMBER output file we'll reintroduce the possibility to extract strings, maybe with an optional argument to the method, like `extract_section(return_str=True)`.

I'm not so worried to have an incorrectly formatted file as this is the main AMBER output, it's pretty stable as it has been formatted the same for more than 10 years! But it's indeed a complication we can avoid for now, as the options that are indeed `str` are not useful for our purposes!